Make type-hint tag (#1943)
python-discord/bot: bot/resources/tags/type-hint.md

**Type Hints**

> A typehint indicates what type something should be. For example,

Reviewer: ```suggestion
A typehint indicates what type something is expected to be. For example,
```

Would this be an improvement in wording?

Reviewer 2: "Type hint(s)" should be two separate words

Reviewer 3: yes

```python
def add(a: int, b: int) -> int:
    sum: int = a + b
    return sum
```

> In this case, `a` and `b` are expected to be ints, and the function returns an int. We also declare an intermediate variable `sum`, which we indicate to be an int.

Reviewer: Should probably mention that the `sum` variable can be inferred (or guessed could be a easier word to understand) by whatever static type checker the user is using.

It's important to note these are just hints and have no runtime effect. For example,
```python

> #uh oh

Reviewer: ```suggestion
# Uh oh!
```

add("hello ", "world")
```

> This code will run without error, even though it doesn't follow the function's type hints.

Reviewer: ```suggestion
This code will run without error because strings can also be added together, even though it doesn't follow the function's type hints.
```

Something like this or similar, we should make it clear that it works even though it's the wrong type because the type we passed in allows the same operations - a kind of gateway into duck typing.

**Type Hints**

A type hint indicates what type something is expected to be. For example,
```python
def add(a: int, b: int) -> int:
    return a + b
```

> In this case, we have a function,`add`, with parameters `a` and `b`. The type hints indicate that the parameters and return type are all integers.

Reviewer 4: Missing a space here after the comma

It's important to note these are just hints and have no runtime effect. For example,
```python
# Uh oh
add("hello ", "world")
```

> This code won't error even though it doesn't follow the function's type hints. It will just concatenate the two strings.

Reviewer 4: I'd add a comma here after the "error" -- "...won't error, even though..."

Reviewer 5: ^

> Third party tools like [mypy](http://mypy-lang.org/) can enforce your type hints. Mypy would error in the second example.

Reviewer 4: Picky but should probably be `MyPy` not `mypy`. Same for after the period.

Reviewer 5: Also feels worth linking pyright and other checkers.

Reviewer 6: I changed to link the docs. I'm not sure mentioning pyright adds much here, and a link I added mentions other type chackers.

> For more info about type hints, check out [PEP 484](https://www.python.org/dev/peps/pep-0484/).

Reviewer 4: Not sure if we want a newline before this or not. I think it's better without it.

> For more info about type hints, check out [PEP 484](https://www.python.org/dev/peps/pep-0484/).

Reviewer 7: Not sure if this would be out of scope for this tag but I think it'd be a nice addition to reference https://typing.readthedocs.io/en/latest/index.html as a more general user focused resource on static typing as it'd be next logical step after learning about type-hints IMO.

Reviewer 8: Yes, there are plans for adding more content like that but it's limited by people's time.

Reviewer 6: Included it, I also mentioned the `mypy` docs as I think that's probably one of the best places to go for an introduction to type hinting.

> [`mypy`s documentation](https://mypy.readthedocs.io/en/stable/builtin_types.html) contains useful information on type hinting, and for more information check out [this documentation page](https://typing.readthedocs.io/en/latest/index.html).

Reviewer 9: ```suggestion
[mypy's documentation](https://mypy.readthedocs.io/en/stable/builtin_types.html) contains useful information on type hinting, and for more information check out [this documentation page](https://typing.readthedocs.io/en/latest/index.html).
```

It's not in a code block on the line above, so shouldn't be here either. Apostrophe to indicate ownership.
